src/Storages/Kafka/KafkaBlockInputStream.cpp

```diff
@@ -186,14 +186,14 @@ Block KafkaBlockInputStream::readImpl()
 
             total_rows = total_rows + new_rows;
         }
-        else if (buffer->isStalled())
-        {
-            ++failed_poll_attempts;
-        }
         else if (buffer->polledDataUnusable())
         {
             break;
         }
+        else if (buffer->isStalled())
+        {
+            ++failed_poll_attempts;
+        }
```

Comment:
So, reordering these blocks fixes the segfault? Is it possible to add an explanation why this order matters and why it must be this way and not some another way?

Reply:
Not really, the SIGSEGV is fixed by avoiding cleaning messages while parsing (i.e. removing resetIfStopped from ReadBufferFromKafkaConsumer::nextImpl).

This reordering is to avoid extra polling if stop flag isset, i.e. prefer break over incrementing failed poll attempts

Reply:
Maybe we can add some comment there, then? The reason why I'm asking is that this kind of fix is always suspicious: one line removed, segfault is fixed, but there was no comment justifying why it was there, and there is no comment justifying why it must not be there. It made sense to add this line before, maybe we should write down why it was a bad idea, for the future.

Of course, it might be difficult to write a comment for line that does not exist, so if a useful comment doesn't come to mind, let's just leave it as is.

Can we add an integration test for this?

Reply:
Here is the fix for segfault and explanation: https://github.com/ClickHouse/ClickHouse/pull/12075#issuecomment-652630127

Reproduce in the integration test reliable is almost impossible (DROP TABLE should happen in a very specific moment), there are quite a lot of tests doing table drops during consumption, they did not catch the bug. 

It may be possible in the unit test, but it would require mocking a lot of stuff, and the required effort is too big. 

Moving condition lower - is was just a logical error which was leading to a few unneeded retries instead of leaving the loop immediately. (unrelated to the segfault).

Reply:
>Maybe we can add some comment there, then? The reason why I'm asking is that this kind of fix is always suspicious: one line removed, segfault is fixed, but there was no comment justifying why it was there, and there is no comment justifying why it must not be there. It made sense to add this line before

@akuzm Agree, such fixes are always very suspicious.
Comments does not always helps to avoid with re-introducing bugs, since something similar can be added to some helper and/or not always comments will be read.
I think that `assert`/`LOGICAL_ERROR` will be a better idea, will add

>Can we add an integration test for this?

Like @filimonov told, it will be racy

Reply:
Rebased and splitted into separate patches